### PR TITLE
added nodejs code generator

### PR DESCRIPTION
**Pull request by bilalyasar**

### CODE REVIEW

File: hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/Lang.java [hunk 1]
```diff
@@ -40,6 +40,11 @@
      *PYTHON lang
      */
     PY,
+    
+    /**
+     *PYTHON lang
```
Comment (gAmUssA) [inline hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/Lang.java]: @bilalyasar you have duplicate PYTHON lang.
Also, personally I don't like `NODE` as language. `JS` or `JavaScript` would be better.

Comment (gAmUssA) [inline hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/Lang.java]: or even `TS` or `TYPESCRIPT`

File: hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodecCodeGenerator.java [hunk 2]
```diff
@@ -314,6 +314,8 @@ public void generateCodec(CodecModel codecModel, Template codecTemplate) {
             String fileName = codecModel.getClassName() + "." + codecModel.getLang().name().toLowerCase();
             if (codecModel.getLang() == Lang.PY) {
                 fileName = fileName.replaceAll("(.)(\\p{Upper})", "$1_$2").toLowerCase();
+            } else if (codecModel.getLang() == Lang.NODE) {
```
Comment (gAmUssA) [inline hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodecCodeGenerator.java]: @bilalyasar matter of taste and preference, but I prefer `switch` instead of `if - else-if` statements

File: hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodeGenerationUtils.java [hunk 3]
```diff
@@ -56,6 +56,30 @@
         put("com.hazelcast.map.impl.SimpleEntryView", "SimpleEntryView");
     }};
 
+    private static final Map<String, String> JAVA_TO_NODE_TYPES = new HashMap<String, String>() {{
```
Comment (gAmUssA) [inline hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodeGenerationUtils.java]: @bilalyasar inconsistency, `JAVA_TO_NODE_TYPES`

File: hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodeGenerationUtils.java [hunk 4]
```diff
@@ -56,6 +56,30 @@
...
+        put("java.lang.String", "string");
+        put("java.lang.Integer", "int32");
+        put("boolean", "boolean");
+        put("int", "int32");
+        put("com.hazelcast.nio.Address", "Address");
+        put("java.util.List", "list");
+        put("java.util.Set", "set");
+    }};
+
+    private static final Map<String, String> JAVA_TO_TS_TYPES = new HashMap<String, String>() {{
```
Comment (gAmUssA) [inline hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodeGenerationUtils.java]: @bilalyasar inconsistency, `JAVA_TO_TS_TYPES`. We need to stick either `TS` or `JS`. `TS` makes more sense because we're producing typescript output

File: hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodeGenerationUtils.java [hunk 5]
```diff
@@ -56,6 +56,30 @@
...
     }};
 
+    private static final Map<String, String> JAVA_TO_NODE_TYPES = new HashMap<String, String>() {{
+        put(DATA_FULL_NAME, "data");
+        put("java.lang.String", "string");
+        put("java.lang.Integer", "int32");
+        put("boolean", "boolean");
+        put("int", "int32");
+        put("com.hazelcast.nio.Address", "Address");
+        put("java.util.List", "list");
```
Comment (mustafaiman) [inline hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodeGenerationUtils.java]: We use Arrays for lists. I am not sure how to represent an array here. Because we create a new array as `var a: Array<type> = [];`.

File: hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodeGenerationUtils.java [hunk 6]
```diff
@@ -56,6 +56,30 @@
...
+        put("java.util.Set", "set");
```
Comment (mustafaiman) [inline hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodeGenerationUtils.java]: Should be `Set`.

File: hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodeGenerationUtils.java (hunk 4 above)
Comment (mustafaiman) [inline hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodeGenerationUtils.java]: @bilalyasar Why do we have two maps `JAVA_TO_NODE_TYPES` and `JAVA_TO_TS_TYPES`? Should not it be either one of them?

File: hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/Lang.java (hunk 1 above)
Comment (mustafaiman) [inline hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/Lang.java]: I think it should be `TYPESCRIPT`.

File: hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodeGenerationUtils.java (hunk 3 above)
Comment (bilalyasar) [inline hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodeGenerationUtils.java]: actually, this parameters are used for method calls. Like read`Int32` or read`Boolean`...

File: hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodeGenerationUtils.java (hunk 6 above)
Comment (bilalyasar) [inline hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodeGenerationUtils.java]: it automatically capitalises first character.

File: hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodeGenerationUtils.java (hunk 5 above)
Comment (bilalyasar) [inline hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodeGenerationUtils.java]: this parameters are not related with types. They are for method declarations.

File: hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodeGenerationUtils.java (hunk 4 above)
Comment (bilalyasar) [inline hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodeGenerationUtils.java]: because one for method declarations, one for ts types. They are different, so i need to create 2 mapping.

File: hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/Lang.java (hunk 1 above)
Comment (bilalyasar) [inline hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/Lang.java]: :+1:

File: hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodecCodeGenerator.java (hunk 2 above)
Comment (bilalyasar) [inline hazelcast-code-generator/src/main/java/com/hazelcast/client/protocol/generator/CodecCodeGenerator.java]: :+1:

**Comment by mustafaiman** [comment]
:+1:

**Comment by mustafaiman** [comment]
:+1:
